utilities/imgls.pl

```diff
@@ -0,0 +1,562 @@
+my $def_preserve_ratio  = 'true';               # preserve aspect ratio
+
+my %imgparams = (                               # default image option parameters
+    inline              => 1,                   # image appears inline with text
+    height              => $def_image_height,   # character cells tall
+    width               => $def_image_width,    # character cells wide
+    preserveAspectRatio => $def_preserve_ratio, # no stretchmarks please
+);
+
+my %failed_types;                               # cache of file extensions that have failed
+my %stat_cache;                                 # cache of file/directory lstat() calls
+my $curtime = time();
+my $sixmonths = (365 / 2) * 86400;
+
+my %opts = (
+    height      => \$imgparams{'height'},
+    width       => \$imgparams{'width'},
+);
+get_options(\%opts);
+
+# find a method to obtain image dimensions
+my $dims_methods = init_dims_methods();
+my $dims_method  = find_dims_methods($dims_methods);
+
+# single pixel image for non-renderable files
+my ($one_pixel_black, $one_pixel_black_len) = get_black_pixel_image();
+
+my $do_newline;
+my $dot_filter = $opts{'A'} ? qr/^\.{1,2}$/ : qr/^\./;
+$dot_filter    = undef          if $opts{'a'};
+
+# special: empty @ARGV, or contains only '.'
+my $do_header = @ARGV > 1 ? 1 : 0;
+if (@ARGV <= 1) {
+    push @ARGV, '.'     if @ARGV == 0;
+}
+
+my (@files, @dirs);
+for (@ARGV) {
+    if (! -e _lstat($_)) {
+        say "$prog: $_: No such file or directory";
+    }
+    elsif (-f _lstat($_)) {
+        push @files, $_;
+    }
+    else {
+        push @dirs, $_;
+    }
+}
+@files = ls_sort(@files);
+@dirs  = ls_sort(@dirs);
+
+if ($opts{'d'}) {
+    push @files, @dirs;
+    @dirs = ();
+}
+do_ls(undef, @files)    if @files;
+
+while (@dirs) {
+    my $path = shift @dirs;
+
+    if (! -e $path) {
+        say "$prog: $path: No such file or directory";
+        next;
+    }
+    my (@f, @d);
+    get_dir_content($path, $dot_filter, \@f, \@d) or
+        next;
+    do_ls($path, @f, @d);
+    if ($opts{'R'}) {
+        push @dirs, grep { ! /\.\.?$/ } @d;
+    }
+    $do_newline++;
+}
+
+# Encodes and outputs an image file to the window
+#    arg 1: path to an image file
+#    arg 2: size, in bytes, of the image
+# Image data is written to STDERR so filters on STDOUT work as expected.
+sub write_image {
+    my ($file, $size) = @_;
+    my $encoded;
+
+    $imgparams{'name'} = encode_base64($file, '');                      # file name is base64 encoded
+    $imgparams{'size'} = $size;                                         # file size in bytes
+
+    if (ref $file eq '')  {
+        my $bytes = get_image_bytes($file);
+        $encoded = encode_base64($bytes)                if defined $bytes;
+    }
+    if (! $encoded or ref $file eq 'SCALAR') {
+        $encoded = $one_pixel_black;
+        $imgparams{'name'} = encode_base64('one_pixel_black', '');
+        $imgparams{'size'} = $one_pixel_black_len;
+    }
+
+    printf STDERR "%s%s%s;File=%s:%s%s",
```

Comment:
Since this is regular output it should go to STDOUT, not STDERR

Reply:
I'll change this.  There's a comment above about being able placing the image on STDOUT; it prevents filters from working.